File: scylla-server-go/services/datatype_service/datatype_service.go
package datatype_service

import (
	"errors"
	"log"
	"scylla-server/prisma"
	"scylla-server/prisma/db"
)

// Get all of the datatypes as a list, returns empty if none exist
//
// Returns nil datatypes and the error if the db errors
func GetAllDataTypes() ([]db.DataTypeModel, error) {
	client, ctx := prisma.PrismaClient()

	data, err := client.DataType.FindMany().Exec(ctx)
	if err != nil {
		if errors.Is(err, db.ErrNotFound) {
			return []db.DataTypeModel{}, nil
		} else {
			log.Printf("Err %q\n", err.Error())
			return nil, err
		}
	}

	return data, nil
}

// Upserts the data type, returning it or the error
func UpsertDataType(dataTypeName string, unit string, nodeName string) (*db.DataTypeModel, error) {
	client, ctx := prisma.PrismaClient()

	data, err := client.DataType.UpsertOne(db.DataType.Name.Equals(dataTypeName)).Create(db.DataType.Name.Set(dataTypeName), db.DataType.Unit.Set(unit), db.DataType.Node.Link(db.Node.Name.Equals(nodeName))).Update(db.DataType.Name.Set(dataTypeName), db.DataType.Unit.Set(unit), db.DataType.Node.Link(db.Node.Name.Equals(nodeName))).Exec(ctx)
	if err != nil {
		log.Printf("Err %q\n", err.Error())
		return nil, err
	}
	return data, nil
}

File: scylla-server-rust/src/services/data_type_service.rs
use prisma_client_rust::QueryError;

use crate::{prisma, Database};

prisma::data_type::select! {public_datatype {
    name
    unit
}}

/// Gets all datatypes
/// * `db` - The prisma client to make the call to
/// returns: A result containing the data or the QueryError propogated by the db
pub async fn get_all_data_types(db: &Database) -> Result<Vec<public_datatype::Data>, QueryError> {
    db.data_type()
        .find_many(vec![])
        .select(public_datatype::select())
        .exec()
        .await
}

/// Upserts a datatype, either creating or updating one depending on its existence
/// * `db` - The prisma client to make the call to
/// * `data_type_name` - The data type name to upsert
/// * `unit` - The unit of the data
/// * `node_name` - The name of the node linked to the data type, must already exist!
/// returns: A result containing the data or the QueryError propogated by the db
pub async fn upsert_data_type(
    db: &Database,
    data_type_name: String,
    unit: String,
    node_name: String,
) -> Result<public_datatype::Data, QueryError> {
    db.data_type()
        .upsert(
            prisma::data_type::name::equals(data_type_name.clone()),
            prisma::data_type::create(
                data_type_name.clone(),
                unit.clone(),
                prisma::node::name::equals(node_name.clone()),
                vec![],
            ),
            vec![
                prisma::data_type::unit::set(unit),
                prisma::data_type::name::set(data_type_name.clone()),
                prisma::data_type::node::connect(prisma::node::name::equals(node_name.clone())),
            ],
        )
        .select(public_datatype::select())
        .exec()
        .await
}
